**Linux** [Build 8661](http://brontes.lab.eng.brq.redhat.com/viewLog.html?buildId=76671&buildTypeId=WFPR) outcome was **FAILURE** using a merge of a6fe290198d4978ef35d17b9e438730b9efca795
Summary: Tests failed: 1 (1 new), passed: 3282, ignored: 363 Build time: 00:51:09
### Failed tests
```
org.jboss.as.test.integration.ejb.mdb.ejb2x.MDB20TopicTestCase.testEjb20TopicMDBs: java.lang.AssertionError: Reply message was null on reply queue: ActiveMQQueue[ejb2x/replyQueueA]
	at org.jboss.as.test.integration.ejb.mdb.ejb2x.MDB20TopicTestCase.testEjb20TopicMDBs(MDB20TopicTestCase.java:117)
------- Stdout: -------
 [0m14:00:37,257 INFO  [org.jboss.as.server.deployment] (MSC service thread 1-2) WFLYSRV0028: Stopped deployment mdb.jar (runtime-name: mdb.jar) in 14ms
 [0m [0m14:00:37,262 INFO  [org.jboss.as.repository] (management-handler-thread - 3) WFLYDR0002: Content removed from location /opt/buildAgent/work/a31d203e70e89f90/testsuite/integration/basic/target/jbossas/standalone/data/content/39/3aead88ac6d95030e095808d218e265b5a60ae/content
 [0m [0m14:00:37,262 INFO  [org.jboss.as.server] (management-handler-thread - 3) WFLYSRV0009: Undeployed "mdb.jar" (runtime-name: "mdb.jar")
 [0m14:00:37,287 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "remove",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "ejb2x/queue")
    ]
}
14:00:37,295 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "remove",
    "address" => [
        ("subsystem" => "messaging-activemq"),
        ("server" => "default"),
        ("jms-queue" => "ejb2x/replyQueue")
    ]
}
14:00:37,301 INFO  [org.jboss.as.test.integration.ejb.mdb.ejb2x.AbstractMDB2xTestCase] (main) MDB20TopicTestCase.jar:
/org/
/org/jboss/
/org/jboss/as/
/org/jboss/as/test/
/org/jboss/as/test/integration/
/org/jboss/as/test/integration/ejb/
/org/jboss/as/test/integration/ejb/mdb/
/org/jboss/as/test/integration/ejb/mdb/ejb2x/
/org/jboss/as/test/integration/ejb/mdb/ejb2x/MDB20TopicTestCase$JmsQueueSetup.class
/org/jboss/as/test/integration/ejb/mdb/ejb2x/EJB2xMDB.class
/org/jboss/as/test/integration/ejb/mdb/ejb2x/AbstractMDB2xTestCase.class
/org/jboss/as/test/integration/common/
/org/jboss/as/test/integration/common/jms/
/org/jboss/as/test/integration/common/jms/JMSOperations.class
/org/jboss/as/test/integration/common/jms/JMSOperationsProvider.class
/org/jboss/as/test/integration/common/jms/ActiveMQProviderJMSOperations.class
/org/jboss/as/test/integration/common/jms/JMSOperationsException.class
/org/jboss/as/test/shared/
/org/jboss/as/test/shared/TimeoutUtil.class
/META-INF/
/META-INF/MANIFEST.MF
/META-INF/ejb-jar.xml
/META-INF/jboss-ejb3.xml
 [0m14:00:37,311 INFO  [org.apache.activemq.artemis.core.server] (ServerService Thread Pool -- 27) AMQ221003: trying to deploy queue jms.topic.ejb2x/topic
 [0m14:00:37,363 INFO  [org.jboss.as.test.integration.common.jms.JMSOperationsProvider] (main) Operation successful for update = {
    "operation" => "add",
    "address" => [
        ("subsystem" => "messaging-activemq"),


```

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/injection/WSComponent.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/injection/WSComponent.java
@@ -46,18 +46,21 @@ public final class WSComponent extends BasicComponent {
     }
 
     public BasicComponentInstance getComponentInstance() {
-       if (wsComponentInstance == null) {
-           synchronized (lock) {
-               if (wsComponentInstance == null && reference == null) {
-                   wsComponentInstance = (BasicComponentInstance) createInstance();
-               }
-               if (wsComponentInstance == null && reference != null) {
-                   wsComponentInstance = (BasicComponentInstance) this.createInstance(reference.getInstance());
-               }
-           }
-       }
-       return wsComponentInstance;
-    }
+        BasicComponentInstance result = wsComponentInstance;
+        if (result == null) {
+            synchronized (lock) {
+                result = wsComponentInstance;
+                if (result == null) {
+                    if (reference == null) {
+                        wsComponentInstance = result = (BasicComponentInstance) createInstance();
+                    } else {
+                        wsComponentInstance = result = (BasicComponentInstance) this.createInstance(reference.getInstance());
+                    }
+                }
+            }
+        }
+        return result;
+     }
 
     public void setReference(ManagedReference reference) {
         this.reference = reference;

--- a/webservices/server-integration/src/main/java/org/jboss/as/webservices/invocation/AbstractInvocationHandler.java
+++ b/webservices/server-integration/src/main/java/org/jboss/as/webservices/invocation/AbstractInvocationHandler.java
@@ -50,7 +50,7 @@ import org.jboss.wsf.spi.invocation.Invocation;
  */
 abstract class AbstractInvocationHandler extends org.jboss.ws.common.invocation.AbstractInvocationHandler {
 
-   private ServiceName componentViewName;
+   private volatile ServiceName componentViewName;
    private volatile ComponentView componentView;
    protected volatile ManagedReference reference;
 
@@ -63,33 +63,36 @@ abstract class AbstractInvocationHandler extends org.jboss.ws.common.invocation.
        componentViewName = (ServiceName) endpoint.getProperty(COMPONENT_VIEW_NAME);
    }
 
-   /**
-    * Gets endpoint container lazily.
-    *
-    * @return endpoint container
-    */
-   protected ComponentView getComponentView() {
-       // we need to check both, otherwise it is possible for
-       // componentView to be initialized before reference
-      if (componentView == null) {
-         synchronized(this) {
-            if (componentView == null) {
-               componentView = getMSCService(componentViewName, ComponentView.class);
-               if (componentView == null) {
-                  throw WSLogger.ROOT_LOGGER.cannotFindComponentView(componentViewName);
-               }
-               if (reference == null) {
-                  try {
-                      reference = componentView.createInstance();
-                  } catch (Exception e) {
-                      throw new RuntimeException(e);
-                  }
-               }
+    /**
+     * Gets endpoint container lazily.
+     *
+     * @return endpoint container
+     */
+    protected ComponentView getComponentView() {
+        ComponentView cv = componentView;
+        // we need to check both, otherwise it is possible for
+        // componentView to be initialized before reference
+        if (cv == null) {
+            synchronized (this) {
+                cv = componentView;
+                if (cv == null) {
+                    cv = getMSCService(componentViewName, ComponentView.class);
+                    if (cv == null) {
+                        throw WSLogger.ROOT_LOGGER.cannotFindComponentView(componentViewName);
+                    }
+                    if (reference == null) {
+                        try {
+                            reference = cv.createInstance();
+                        } catch (Exception e) {
+                            throw new RuntimeException(e);
+                        }
+                    }
+                    componentView = cv;
+                }
             }
-         }
-      }
-      return componentView;
-   }
+        }
+        return cv;
+    }
 
    /**
     * Invokes WS endpoint.
@@ -106,7 +109,8 @@ abstract class AbstractInvocationHandler extends org.jboss.ws.common.invocation.
          final ComponentView componentView = getComponentView();
          Component component = componentView.getComponent();
          //in case of @FactoryType annotation we don't need to go into EE interceptors
-         if(wsInvocation.getInvocationContext().getProperty("forceTargetBean") != null) {
+         final boolean forceTargetBean = (wsInvocation.getInvocationContext().getProperty("forceTargetBean") != null);
+         if (forceTargetBean) {
              this.reference = new ManagedReference() {
                  public void release() {
                  }
@@ -126,7 +130,7 @@ abstract class AbstractInvocationHandler extends org.jboss.ws.common.invocation.
          context.setParameters(wsInvocation.getArgs());
          context.putPrivateData(Component.class, component);
          context.putPrivateData(ComponentView.class, componentView);
-         if(wsInvocation.getInvocationContext().getProperty("forceTargetBean") != null) {
+         if(forceTargetBean) {
              context.putPrivateData(ManagedReference.class, reference);
          }
          // invoke method

